### ISSUE TITLE
Paramaterized IPLD nodes / paths

### ISSUE THREAD

**Issue by Gozala** [ipld/ipld#67]
This came up in https://github.com/ipld/ipld/issues/64#issuecomment-473442520 but I think it deserves it's own issue.

Parameterization should have first class support in IPLD so that:

1. Selectors could cross number of encoding layers. 
2. Tools like [IPLD Explorer](https://explore.ipld.io/) can be useful with parametrized nodes.

I would propose to add standardized way for accessing parametrized nodes, specifically I think it would be best to make parameter be part of IPLD path (note that query parameters won't work as IPLD paths may span across several nodes each requiring parameters that may conflict).

While it is not limited to encrypted blocks, that is the easiest way to describe problem - imagine wanting to represent encrypted nodes via e.g`dag-secretbox` encoding that along with `message` needs `nonce` and `secret` to encode / decode a message. In this example node of the `dag-secretbox` has following shape `{message, nonce, secret}` and encoding is lossy, there for decoding requires `(nonce, secret)` in addition to the encoded blob.

However navigating to `dag-secretbox` node `CID` should still be navigable, furthermore it would be great if:

1. [IPLD Explorer](https://explore.ipld.io/) was able to provide fields that correspond to required parameters.
2. It was possible to provide a IPLD path (explorer URL) that provides direct access to the decrypted secret message.

I propose to think of parametric nodes as nodes with opaque / query links. Specifically secret message: `"my little secret"`, encrypted with `"shared-secret"` secret key  and `1234` nonce can be visualized as:

```json
{
  "1234": {
     "shared-secret": "my little secret"
  }
}
```

And there for secret could be seamlessly decrypted when using following path: `/${cid}/1234/shared-secret`.

At the same time "dag-secretbox" would not (obviously) provide `nonce: 1234` and `secret: shared-secret` when traversing that node. Or how could it, actual blob would be encrypted. What "dag-secretbox" would instead want to do is point out that there are two nested opaque query links:

```json
{
  "nonce?": {
    "secret?": message
  }
}
```

In other words think of node that has secret / opaque links contents of which can be accessed if link is known. That is 

I propose to extend IPLD API slightly and assign paths that end with `?` character special meaning, that is query opaque / secret path. This would also allow format / codec to provide additional context to required parameters by providing a name for it. Here is example how mentioned `dag-secretbox` could be implemented:

```js
// Assume promise based API instead of callback base one
const Secretbox = {
  format: "dag-secretbox",
  async serialize({message, nonce, key}) {
    return nacl.secretbox(message, nonce, key)
  },
  async deserialize(box) {
    throw Error(`Can not access secret message, without nonce & secret`)
  },
  async resolve(blob, path) {
    const [root, ...params] = path.split("/")
    switch (params.length) {
      case 0:
        return ["/nonce?/key?"]
      case 1:
        return ["/key?"]
      case 2:
        return {
          value: await nacl.secretbox.open(box, ...params)
        }
      default:
        throw new Error('path out of scope')
    }
  },
  async tree(blob) {
    return ["/nonce?/key?"]
  }
}
```

Would be nice to allow format / codec implementer also provide some type information for the parameter, but that could be a future improvement.

Also note that as proposed it will be up to format / codec implementation to choose which parameters are optional and which are mandatory. For example secretbox could have provided two virtual paths to access secret message `["/nonce+key?", "/nonce?/key?"]` (first in this case would have being just concatenation of `nonce+key`

**Comment by Gozala** [comment]
Note this also would very well work with [private-box](https://github.com/auditdrivencrypto/private-box) used for multi recipient [private messages in ssb](https://ssbc.github.io/scuttlebutt-protocol-guide/#private-messages) where message, recipients and their number is concealed from everyone & only recipient can access message using own private key.

**Comment by Gozala** [comment]
It's also worth pointing out that if inline-node support was added as proposed in https://github.com/ipld/js-ipld-block/issues/47 it would automatically allow multiple layers of encryptions be crossed through paths.

**Comment by mikeal** [comment]
We’re still figuring out all the use cases and trying to build a coherent layer model for IPLD, which I think we desperately need in order to fully consider this proposal.

While it may look today like the block layer and selectors are rather close to each other, I think that in the long term they may not be, and selectors may end up being a much higher level interface than they are today. We’re already talking about ways to make collections (like hamt) self-describing enough that a selector engine will treat the entire tree as a single namespace as far as the selector is concerned.

So, I have a few questions that can start to bring clarity to this proposal and where we should fit it in the layer model we’re working on.

1. Is this something that can be accomplished at a layer above the block format?
2. Are parameters sufficiently self-describing? Can I read any node and know that it requires certain parameters or not? Not only for use cases that we have now, but future ones we haven’t yet considered.
3. It’s clear that parameters would need to be passed into selectors, what are the other read operations and APIs that would also need to support parameters? 
4. Are there low level read operations, like full tree replication (our current pinning API), that **DO NOT** need to be aware of parameters?

**Comment by mikeal** [comment]
Side Note: Can anyone point me at the rationale for CBOR? It was quite a while ago, long before I got involved, and I’ve been able to pick up parts of the rationale but I’m not sure if I have the complete picture.

**Comment by Gozala** [comment]
> So, I have a few questions that can start to bring clarity to this proposal and where we should fit it in the layer model we’re working on.
> 
>     1\. Is this something that can be accomplished at a layer above the block format?
> 

First, let me challenge the block format notion here (or at least my interpretation). I tried to decouple this from the [inline-block](https://github.com/ipld/js-ipld-block/issues/47) proposal I have being also being trying to push, which I think may connect some dots.

At the moment it seems IPLD does tie the memory layout very closely to the logical layout of the data - block encodes data in a specific format and may have links to blocks of data in a different format. I IPLD to truly fulfill it's promise it needs a way to be agnostic of memory layout as in one block may contain inlined blocks in encoded in arbitrary other format.

I believe that could be best accomplished by allowing blocks to have an inline blocks (as in no CIDs, just a slice of prefixed bytes - multicodec). That provides a way to think of blocks as logical / virtual data layout without reflecting it in the actual storage layer.

I suspect that codec / format will be encoding parameter related metadata into blocks as well. But I also feel that codec should be in charge how the parameter information is encoded and interpreted rather than making it format agnostic (could be convinced otherwise). Making it format agnostic would have following trade-offs:

- Some codecs do not need to encode anything into block (like my secret-box example). Encoding that information would be less memory efficient.
- Codecs in the proposed design are free to interpret parameters in arbitrary way. As mentioned secret-box example could look at first param and intepret it either as `nonce` or `nonce+secret` and based on match request second parameter or reveal secret. In other words it can have optional arguments and interpretation of order is also up to it. Encoding all this at the block level would be more complex.

With that all that said, I think you can do it at the different layer, however I'm inclined to think that proposed layer is the right one. As at the end of the day codec need to utilize parameters and it will still need feedback loop when parameters are invalid and you wind up still doing some of the work at the codec layer and also accept trade-offs mentioned.

>
>     2. Are parameters sufficiently self-describing? Can I read any node and know that it requires certain parameters or not? Not only for use cases that we have now, but future ones we haven’t yet considered.

They are not, nor they should be IMO. You need to consult codec / format to make any sense of the node anyway so why not let it to also tell you all about the parameters. That way it's free to encode that information as it makes sense for that data format.

> 
>     3. It’s clear that parameters would need to be passed into selectors, what are the other read operations and APIs that would also need to support parameters?
> 

This is little too abstract, can you unpack this a little bit ? 

Here is the mental model I'd like to invite for thinking about this
Codec / Format is machine that you can put data of some shape into that it will take apart and return a box hosting the pieces (encode operation). Putting box back into the machine will assemble data from the pieces in it and return it. Some machines however may require keys, passwords or some settings to be able to unlock the provided boxes. In fact some machines may even operate on boxes with multiple compartments and depending on keys, settings used may either assemble the data back together or return another nested box that was hosted with in it (access control).

In other words I don't think of it as read OP but more of as a way to interpret the format.

>     4. Are there low level read operations, like full tree replication (our current pinning API), that **DO NOT** need to be aware of parameters?

I'm not sure I understand your point here, but it seems to contradict my goal. I do in fact want to have an ability to make tree non-replaceable without knowing a right parameter (secret key) and only allow replication if corresponding parameter is provided. And if you happen to have another parameter (even more secret key) only then allow you to read the data itself.

**Comment by vmx** [comment]
> Side Note: Can anyone point me at the rationale for CBOR?

grep @daviddias's brain for "Lisbon Treaty"

**Comment by Stebalien** [comment]
> At the moment it seems IPLD does tie the memory layout very closely to the logical layout of the data - block encodes data in a specific format and may have links to blocks of data in a different format. I IPLD to truly fulfill it's promise it needs a way to be agnostic of memory layout as in one block may contain inlined blocks in encoded in arbitrary other format.

> I believe that could be best accomplished by allowing blocks to have an inline blocks (as in no CIDs, just a slice of prefixed bytes - multicodec). That provides a way to think of blocks as logical / virtual data layout without reflecting it in the actual storage layer.

You can already do this by generating a CID of the block you want to inline using the identity hash function. We even do this in the current go-ipfs implementation.

Note: There are very few cases where you should actually do this:

1. When creating a CID to be passed of to the user (i.e., not inlined in another block). When you just need to combine two blocks, you should, when possible, use the same format.
2. When dealing with existing applications/formats. For example, you may want to embed a git commit inside a file's metadata. In this case, you probably care about the specific format.

**Comment by Gozala** [comment]
> You can already do this by generating a CID of the block you want to inline using the identity hash function. We even do this in the current go-ipfs implementation.

Oh that is is neat, I have not realized that *identity* was a multicodec. That seems indeed like a reasonable option.



> Note: There are very few cases where you should actually do this:
> 
>     1. When creating a CID to be passed of to the user (i.e., not inlined in another block). When you just need to combine two blocks, you should, when possible, use the same format.
> 
>     2. When dealing with existing applications/formats. For example, you may want to embed a git commit inside a file's metadata. In this case, you probably care about the specific format.

Use case that has being discussed here does not fit neither of these two, however I would still argue it's a totally valid use case. @Stebalien do you have some feedback / thoughts on that ?

**Comment by vmx** [comment]
> > You can already do this by generating a CID of the block you want to inline using the identity hash function
>
> That seems indeed like a reasonable option.

I don't understand how this would help with inline blocks.

**Comment by Gozala** [comment]
> I don't understand how this would help with inline blocks.

Maybe I've misunderstood something, but I was assuming that identity hash would just return buffer being hashed and there for all of the content will be part of resulting CID.

**Comment by Stebalien** [comment]
@Gozala exactly. The inlined block would just be a very large CID of the form `<cidv1><ipld-codec><identity-multihash-codec><length><data...>`.

**Comment by mikeal** [comment]
@Stebalien as far as the data model is concerned, is this stored as a big binary object or a CID/Link?

**Comment by Stebalien** [comment]
It's stored inside the CID. There are two use-cases:

1. Dealing with the block-boundary issue. This feature came along _before_ the whole "let's just make IPLD block agnostic" resolution.
2. Avoid network requests for tiny files. This is still applicable.

**Comment by mikeal** [comment]
So... none of the code we have in JS considers this case. Our traversal code and the resolve functions in the formats assume that a CID is a *remote* reference that these libraries either return or lookup in a given block store, we have no code (to my knowledge) that accounts for a CID that includes its data pulled from the multihash.

This is a pretty massive oversight and needs to be thought about in a few of our abstractions on the JS side. @warpfork have you take this into account on the Go side?

**Comment by Stebalien** [comment]
We only recently added support for this in go-ipfs. We actually handle it with a special datastore that just extracts the block from the CID. This should be pretty trivial to implement in javascript.

**Comment by Gozala** [comment]
Created https://github.com/ipld/ipld/issues/68 so we can continue discussions on inline / nested blocks there instead of derailing (my fault) this thread.

**Comment by Gozala** [comment]
Few remarks I'd like to note for the reference from my exchange with @Stebalien:

Quoting @Stebalien:
> gozala: So, there's some discussion (https://github.com/ipld/specs/issues/83) for changing IPLD links to paths instead of just CIDs. That would allow one to have some node A with a pointer to `/ipld/CidOfB/nonce:secret/....`
> One concern here is security WRT graphsync as graphsync would send this path to all connected peers.
> note: this is one of the many reasons I'd like to eventually embed a programming system inside IPLD (or some system built on-top of IPLD). One would be able to write `A.lookup("b").decrypt(secret).lookup("c")`.

I agree I don't want to be passing around secrets via paths to the random peers, I do however want to be able to have a way to have paths with secrets so graph could be navigated via [IPLD Explorer](https://explore.ipld.io/). It is however important to consider that this could create a risk of accidental secret leakage and considering ways to mitigate that.

I also want to point out that I'm sort of proposing is a very limited programmable interface for IPLD. Limited because it doesn't require defining language agnostic RPC / Type System nor universal runtime like WASM. Not because I don't think they're bad idea, but rather because I think those are non-trivial problems and have significant trade-offs. Instead I'm proposing to allow codec / format provide path that can be thought of as method / function that accepts single parameter and leave everything else to the codec implementation:
- Cases with multiple parameters can be thought as curried functions
- Type / parameter information is up to codec / format to encode into a block.
- Type / parameter is also up to codec / format to interpret.

I suspect over time this would lead to common patterns that eventually could be standardized.

**Comment by Gozala** [comment]
I also want to point out something that was probably unclear (as it was only in my mind) that I expect that paths would not necessarily contain secrets (shared keys) but rather identifiers to them. More specifically I imagine that shared key will be in the key-chain (of sorts) and giving access to a peer would involve key exchange. Then path to my secret would not need to actually embed the shared key but rather reference it from key-chain via some identifier. In fact typing shared key into IPLD Explorer should probably also import the key into keychain and use identifier in the path rather than embed the key itself.

**Comment by Stebalien** [comment]
> I suspect over time this would lead to common patterns that eventually could be standardized.

What I _don't_ want is a proliferation of new custom formats every time someone wants to define a new function. Every implementation will need to implement every function.

I _also_ don't want to end up with a stringly typed programming language built out of pre-made "formats". That's why the whole type-system thing is going to take a non-trivial amount of effort.

---

To copy down some other pieces of that discussion, this would also fundamentally change (not necessarily a bad thing) the design of IPLD. IPLD was originally defined defined in terms of two functions: `tree` and `resolve`. This change would mean that `tree` isn't implemented on some nodes.

Really, this might not be all that bad. I know that @nicola wants accumulators in IPLD which are, by nature, non-enumerable sets.

> More specifically I imagine that shared key will be in the key-chain (of sorts) and giving access to a peer would involve key exchange. Then path to my secret would not need to actually embed the shared key but rather reference it from key-chain via some identifier. In fact typing shared key into IPLD Explorer should probably also import the key into keychain and use identifier in the path rather than embed the key itself.

That's going to be tricky to get right from a security perspective. Ambient authority is insanely dangerous. We don't want untrusted objects using keys they wouldn't otherwise have access to, that's why I like explicitly putting them in the path (I have to know the secret to use it).

**Comment by tobowers** [comment]
> Side Note: Can anyone point me at the rationale for CBOR? It was quite a while ago, long before I got involved, and I’ve been able to pick up parts of the rationale but I’m not sure if I have the complete picture.

I'm not an IPLD developer, just a user so I can't tell you what the original rationale was, but I can say I've been trying to find other canonical binary formats and I haven't found anything better (or much at all).

**Comment by mikeal** [comment]
Ya, after spending time with CBOR I’m a fan as well, although I also think we need to continue to work on `dag-json` as it has some benefits when shipping code to browsers. 

I have a good idea of the rationale for moving away from `dag-pb` and away from links being triples, which is great and I’m super happy about, but the rationale for that shift along with the move to CBOR are sort of intertwined and I’m still unclear on why CBOR was picked as a format over, say, msgpack.

**Comment by Stebalien** [comment]
I believe the answer is: we had to pick something and CBOR checks all the boxes. CBOR was heavily inspired by msgpack.

**Comment by tobowers** [comment]
json/msgpack/protobuf don't support canonical representation though.

**Comment by warpfork** [comment]
Msgpack has some bonkerinos stuff about being unable to tell the difference between strings and bytes, iirc.  This may not be true anymore -- I think it was eventually patched into a saner state (though not necessarily with an unambiguous migration path) -- but the ecosystemic uncertainty damage was done.

(So, yeah, "not canonical representation".)
